// nestjs-grpc/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcErrorCode = exports.VALIDATION_LIMITS = exports.DEFAULT_RETRY_DELAY = exports.DEFAULT_RETRY_ATTEMPTS = exports.DEFAULT_TIMEOUT = exports.DEFAULT_MAX_MESSAGE_SIZE = exports.GRPC_CLIENT_TOKEN_PREFIX = exports.GRPC_SERVICE_METADATA = exports.GRPC_METHOD_METADATA = exports.GRPC_CONTROLLER_METADATA = exports.GRPC_OPTIONS = void 0;
exports.GRPC_OPTIONS = 'GRPC_OPTIONS';
exports.GRPC_CONTROLLER_METADATA = 'GRPC_CONTROLLER_METADATA';
exports.GRPC_METHOD_METADATA = 'GRPC_METHOD_METADATA';
exports.GRPC_SERVICE_METADATA = 'GRPC_SERVICE_METADATA';
exports.GRPC_CLIENT_TOKEN_PREFIX = 'GRPC_CLIENT_';
exports.DEFAULT_MAX_MESSAGE_SIZE = 4 * 1024 * 1024;
exports.DEFAULT_TIMEOUT = 30000;
exports.DEFAULT_RETRY_ATTEMPTS = 3;
exports.DEFAULT_RETRY_DELAY = 1000;
exports.VALIDATION_LIMITS = {
    MAX_MESSAGE_SIZE: 100 * 1024 * 1024,
    MIN_MESSAGE_SIZE: 1024,
    MAX_TIMEOUT: 5 * 60 * 1000,
    MIN_TIMEOUT: 1000,
    MAX_RETRIES: 10,
    MIN_RETRIES: 0,
    MAX_RETRY_DELAY: 10000,
    MIN_RETRY_DELAY: 100,
};
var GrpcErrorCode;
(function (GrpcErrorCode) {
    GrpcErrorCode[GrpcErrorCode["OK"] = 0] = "OK";
    GrpcErrorCode[GrpcErrorCode["CANCELLED"] = 1] = "CANCELLED";
    GrpcErrorCode[GrpcErrorCode["UNKNOWN"] = 2] = "UNKNOWN";
    GrpcErrorCode[GrpcErrorCode["INVALID_ARGUMENT"] = 3] = "INVALID_ARGUMENT";
    GrpcErrorCode[GrpcErrorCode["DEADLINE_EXCEEDED"] = 4] = "DEADLINE_EXCEEDED";
    GrpcErrorCode[GrpcErrorCode["NOT_FOUND"] = 5] = "NOT_FOUND";
    GrpcErrorCode[GrpcErrorCode["ALREADY_EXISTS"] = 6] = "ALREADY_EXISTS";
    GrpcErrorCode[GrpcErrorCode["PERMISSION_DENIED"] = 7] = "PERMISSION_DENIED";
    GrpcErrorCode[GrpcErrorCode["RESOURCE_EXHAUSTED"] = 8] = "RESOURCE_EXHAUSTED";
    GrpcErrorCode[GrpcErrorCode["FAILED_PRECONDITION"] = 9] = "FAILED_PRECONDITION";
    GrpcErrorCode[GrpcErrorCode["ABORTED"] = 10] = "ABORTED";
    GrpcErrorCode[GrpcErrorCode["OUT_OF_RANGE"] = 11] = "OUT_OF_RANGE";
    GrpcErrorCode[GrpcErrorCode["UNIMPLEMENTED"] = 12] = "UNIMPLEMENTED";
    GrpcErrorCode[GrpcErrorCode["INTERNAL"] = 13] = "INTERNAL";
    GrpcErrorCode[GrpcErrorCode["UNAVAILABLE"] = 14] = "UNAVAILABLE";
    GrpcErrorCode[GrpcErrorCode["DATA_LOSS"] = 15] = "DATA_LOSS";
    GrpcErrorCode[GrpcErrorCode["UNAUTHENTICATED"] = 16] = "UNAUTHENTICATED";
})(GrpcErrorCode || (exports.GrpcErrorCode = GrpcErrorCode = {}));
