// nestjs-grpc/dist/exceptions/grpc.exception-filter.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcExceptionFilter = void 0;
const common_1 = require("@nestjs/common");
const rxjs_1 = require("rxjs");
const logger_1 = require("../utils/logger");
const grpc_exception_1 = require("./grpc.exception");
let GrpcExceptionFilter = class GrpcExceptionFilter {
    constructor(options = {}) {
        this.options = {
            enableLogging: options.enableLogging ?? true,
            maxMessageLength: options.maxMessageLength ?? 1000,
            fallbackMessage: options.fallbackMessage ?? 'Internal server error occurred',
            fallbackCode: options.fallbackCode ?? 13,
        };
        this.logger = new logger_1.GrpcLogger({
            enabled: this.options.enableLogging,
            level: 'error',
            context: 'GrpcExceptionFilter',
        });
    }
    catch(exception, _host) {
        try {
            const error = this.normalizeError(exception);
            if (this.options.enableLogging) {
                this.logError(error, exception);
            }
            return (0, rxjs_1.throwError)(() => error);
        }
        catch (filterError) {
            const fallbackError = {
                code: this.options.fallbackCode,
                message: this.options.fallbackMessage,
                details: { originalError: String(exception) },
            };
            if (this.options.enableLogging) {
                this.logger.error('Exception filter failed to process error', filterError);
                this.logger.error('Original exception', exception);
            }
            return (0, rxjs_1.throwError)(() => fallbackError);
        }
    }
    normalizeError(exception) {
        if (exception instanceof grpc_exception_1.GrpcException) {
            return {
                code: exception.getCode(),
                message: exception.message,
                details: exception.getDetails(),
                metadata: exception.getMetadata(),
            };
        }
        if (exception && typeof exception.getError === 'function') {
            const error = exception.getError();
            if (error && typeof error === 'object') {
                return {
                    code: error.code ?? 13,
                    message: error.message ?? exception.message ?? 'Internal server error',
                    details: error.details,
                    metadata: error.metadata,
                };
            }
        }
        if (exception instanceof Error) {
            return {
                code: 13,
                message: this.truncateMessage(exception.message),
                details: {
                    name: exception.name,
                    stack: exception.stack,
                },
            };
        }
        return {
            code: 13,
            message: this.truncateMessage(typeof exception === 'string' ? exception : 'Unknown error occurred'),
            details: {
                originalError: exception,
            },
        };
    }
    logError(error, _originalException) {
        try {
            const errorMessage = `gRPC error: ${error.message}`;
            const errorDetails = `Code: ${error.code}`;
            this.logger.error(errorMessage, errorDetails);
            if (error.details && typeof error.details === 'object') {
                try {
                    const detailsJson = JSON.stringify(error.details);
                    if (detailsJson.length <= this.options.maxMessageLength) {
                        this.logger.error('Error details', detailsJson);
                    }
                }
                catch {
                }
            }
        }
        catch {
        }
    }
    truncateMessage(message) {
        if (!message || typeof message !== 'string') {
            return 'Unknown error';
        }
        const trimmed = message.trim();
        return trimmed.length > this.options.maxMessageLength
            ? `${trimmed.substring(0, this.options.maxMessageLength)}...`
            : trimmed;
    }
};
exports.GrpcExceptionFilter = GrpcExceptionFilter;
exports.GrpcExceptionFilter = GrpcExceptionFilter = __decorate([
    (0, common_1.Catch)(),
    __metadata("design:paramtypes", [Object])
], GrpcExceptionFilter);
