// nestjs-grpc/dist/grpc.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var GrpcModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcModule = void 0;
require("reflect-metadata");
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const constants_1 = require("./constants");
const grpc_exception_1 = require("./exceptions/grpc.exception");
const grpc_client_service_1 = require("./services/grpc-client.service");
const grpc_controller_discovery_service_1 = require("./services/grpc-controller-discovery.service");
const grpc_proto_service_1 = require("./services/grpc-proto.service");
const grpc_provider_service_1 = require("./services/grpc-provider.service");
const grpc_registry_service_1 = require("./services/grpc-registry.service");
let GrpcModule = GrpcModule_1 = class GrpcModule {
    static forProvider(options) {
        this.validateGrpcOptions(options);
        const providers = [
            {
                provide: constants_1.GRPC_OPTIONS,
                useValue: options,
            },
            grpc_proto_service_1.GrpcProtoService,
            grpc_client_service_1.GrpcClientService,
            grpc_provider_service_1.GrpcProviderService,
            grpc_registry_service_1.GrpcRegistryService,
            grpc_controller_discovery_service_1.GrpcControllerDiscoveryService,
        ];
        return {
            module: GrpcModule_1,
            global: true,
            imports: [core_1.DiscoveryModule],
            providers,
            exports: [
                constants_1.GRPC_OPTIONS,
                grpc_proto_service_1.GrpcProtoService,
                grpc_client_service_1.GrpcClientService,
                grpc_provider_service_1.GrpcProviderService,
                grpc_registry_service_1.GrpcRegistryService,
                grpc_controller_discovery_service_1.GrpcControllerDiscoveryService,
            ],
        };
    }
    static forProviderAsync(options) {
        if (!options || typeof options !== 'object') {
            throw new Error('Async options must be a valid object');
        }
        const providers = [
            ...this.createAsyncProviders(options),
            ...(options.providers ?? []),
            grpc_proto_service_1.GrpcProtoService,
            grpc_client_service_1.GrpcClientService,
            grpc_provider_service_1.GrpcProviderService,
            grpc_registry_service_1.GrpcRegistryService,
            grpc_controller_discovery_service_1.GrpcControllerDiscoveryService,
        ];
        return {
            module: GrpcModule_1,
            global: true,
            imports: [core_1.DiscoveryModule, ...(options.imports ?? [])],
            providers,
            exports: [
                constants_1.GRPC_OPTIONS,
                grpc_proto_service_1.GrpcProtoService,
                grpc_client_service_1.GrpcClientService,
                grpc_provider_service_1.GrpcProviderService,
                grpc_registry_service_1.GrpcRegistryService,
                grpc_controller_discovery_service_1.GrpcControllerDiscoveryService,
            ],
        };
    }
    static forConsumer(options) {
        this.validateConsumerOptions(options);
        const providers = [
            {
                provide: constants_1.GRPC_OPTIONS,
                useValue: {
                    protoPath: options.protoPath,
                    package: options.package,
                    url: options.url,
                    secure: options.secure,
                    rootCerts: options.rootCerts,
                    privateKey: options.privateKey,
                    certChain: options.certChain,
                    loaderOptions: options.loaderOptions,
                    logging: options.logging,
                    serviceName: options.serviceName,
                    timeout: options.timeout,
                    maxRetries: options.maxRetries,
                    retryDelay: options.retryDelay,
                    channelOptions: options.channelOptions,
                },
            },
            grpc_proto_service_1.GrpcProtoService,
            grpc_client_service_1.GrpcClientService,
            grpc_exception_1.GrpcConsumerErrorHandler,
        ];
        return {
            module: GrpcModule_1,
            imports: [core_1.DiscoveryModule],
            providers,
            exports: [grpc_client_service_1.GrpcClientService, grpc_exception_1.GrpcConsumerErrorHandler],
        };
    }
    static forConsumerAsync(options) {
        if (!options || typeof options !== 'object') {
            throw new Error('Async consumer options must be a valid object');
        }
        const providers = [
            ...this.createConsumerAsyncProviders(options),
            ...(options.providers ?? []),
            grpc_proto_service_1.GrpcProtoService,
            grpc_client_service_1.GrpcClientService,
            grpc_exception_1.GrpcConsumerErrorHandler,
        ];
        return {
            module: GrpcModule_1,
            imports: [core_1.DiscoveryModule, ...(options.imports ?? [])],
            providers,
            exports: [grpc_client_service_1.GrpcClientService, grpc_exception_1.GrpcConsumerErrorHandler],
        };
    }
    static createAsyncProviders(options) {
        const providers = [];
        if (options.useFactory && typeof options.useFactory === 'function') {
            providers.push({
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (...args) => {
                    const grpcOptions = await options.useFactory(...args);
                    this.validateGrpcOptions(grpcOptions);
                    return grpcOptions;
                },
                inject: options.inject ?? [],
            });
        }
        else if (options.useClass) {
            providers.push({
                provide: options.useClass,
                useClass: options.useClass,
            }, {
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (optionsFactory) => {
                    const grpcOptions = await optionsFactory.createGrpcOptions();
                    this.validateGrpcOptions(grpcOptions);
                    return grpcOptions;
                },
                inject: [options.useClass],
            });
        }
        else if (options.useExisting) {
            providers.push({
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (optionsFactory) => {
                    const grpcOptions = await optionsFactory.createGrpcOptions();
                    this.validateGrpcOptions(grpcOptions);
                    return grpcOptions;
                },
                inject: [options.useExisting],
            });
        }
        else {
            throw new Error('One of useFactory, useClass, or useExisting must be provided');
        }
        return providers;
    }
    static createConsumerAsyncProviders(options) {
        const providers = [];
        if (options.useFactory && typeof options.useFactory === 'function') {
            providers.push({
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (...args) => {
                    const consumerOptions = await options.useFactory(...args);
                    this.validateConsumerOptions(consumerOptions);
                    return {
                        protoPath: consumerOptions.protoPath,
                        package: consumerOptions.package,
                        url: consumerOptions.url,
                        secure: consumerOptions.secure,
                        rootCerts: consumerOptions.rootCerts,
                        privateKey: consumerOptions.privateKey,
                        certChain: consumerOptions.certChain,
                        loaderOptions: consumerOptions.loaderOptions,
                        logging: consumerOptions.logging,
                        serviceName: consumerOptions.serviceName,
                        timeout: consumerOptions.timeout,
                        maxRetries: consumerOptions.maxRetries,
                        retryDelay: consumerOptions.retryDelay,
                        channelOptions: consumerOptions.channelOptions,
                    };
                },
                inject: options.inject ?? [],
            });
        }
        else if (options.useClass) {
            providers.push({
                provide: options.useClass,
                useClass: options.useClass,
            }, {
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (optionsFactory) => {
                    const consumerOptions = await optionsFactory.createGrpcConsumerOptions();
                    this.validateConsumerOptions(consumerOptions);
                    return {
                        protoPath: consumerOptions.protoPath,
                        package: consumerOptions.package,
                        url: consumerOptions.url,
                        secure: consumerOptions.secure,
                        rootCerts: consumerOptions.rootCerts,
                        privateKey: consumerOptions.privateKey,
                        certChain: consumerOptions.certChain,
                        loaderOptions: consumerOptions.loaderOptions,
                        logging: consumerOptions.logging,
                        serviceName: consumerOptions.serviceName,
                        timeout: consumerOptions.timeout,
                        maxRetries: consumerOptions.maxRetries,
                        retryDelay: consumerOptions.retryDelay,
                        channelOptions: consumerOptions.channelOptions,
                    };
                },
                inject: [options.useClass],
            });
        }
        else if (options.useExisting) {
            providers.push({
                provide: constants_1.GRPC_OPTIONS,
                useFactory: async (optionsFactory) => {
                    const consumerOptions = await optionsFactory.createGrpcConsumerOptions();
                    this.validateConsumerOptions(consumerOptions);
                    return {
                        protoPath: consumerOptions.protoPath,
                        package: consumerOptions.package,
                        url: consumerOptions.url,
                        secure: consumerOptions.secure,
                        rootCerts: consumerOptions.rootCerts,
                        privateKey: consumerOptions.privateKey,
                        certChain: consumerOptions.certChain,
                        loaderOptions: consumerOptions.loaderOptions,
                        logging: consumerOptions.logging,
                        serviceName: consumerOptions.serviceName,
                        timeout: consumerOptions.timeout,
                        maxRetries: consumerOptions.maxRetries,
                        retryDelay: consumerOptions.retryDelay,
                        channelOptions: consumerOptions.channelOptions,
                    };
                },
                inject: [options.useExisting],
            });
        }
        else {
            throw new Error('One of useFactory, useClass, or useExisting must be provided');
        }
        return providers;
    }
    static validateConsumerOptions(options) {
        if (!options || typeof options !== 'object') {
            throw new Error('Consumer options must be a valid object');
        }
        if (!options.protoPath || typeof options.protoPath !== 'string') {
            throw new Error('protoPath is required and must be a string');
        }
        if (!options.package || typeof options.package !== 'string') {
            throw new Error('package is required and must be a string');
        }
        if (!options.url || typeof options.url !== 'string') {
            throw new Error('url is required and must be a string');
        }
    }
    static validateGrpcOptions(options) {
        if (!options || typeof options !== 'object') {
            throw new Error('gRPC options must be a valid object');
        }
        if (!options.protoPath || typeof options.protoPath !== 'string') {
            throw new Error('protoPath is required and must be a string');
        }
        if (!options.package || typeof options.package !== 'string') {
            throw new Error('package is required and must be a string');
        }
        if (options.url && typeof options.url !== 'string') {
            throw new Error('url must be a string');
        }
        if (options.maxSendMessageSize !== undefined &&
            (typeof options.maxSendMessageSize !== 'number' || options.maxSendMessageSize <= 0)) {
            throw new Error('maxSendMessageSize must be a positive number');
        }
        if (options.maxReceiveMessageSize !== undefined &&
            (typeof options.maxReceiveMessageSize !== 'number' ||
                options.maxReceiveMessageSize <= 0)) {
            throw new Error('maxReceiveMessageSize must be a positive number');
        }
    }
};
exports.GrpcModule = GrpcModule;
exports.GrpcModule = GrpcModule = GrpcModule_1 = __decorate([
    (0, common_1.Global)(),
    (0, common_1.Module)({
        imports: [core_1.DiscoveryModule],
    })
], GrpcModule);
