// nestjs-grpc/dist/utils/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcLogger = void 0;
const common_1 = require("@nestjs/common");
class GrpcLogger {
    constructor(options = {}) {
        this.options = {
            enabled: options.enabled ?? true,
            level: options.level ?? 'log',
            context: options.context ?? 'GrpcModule',
            logErrors: options.logErrors ?? true,
            logPerformance: options.logPerformance ?? false,
            logDetails: options.logDetails ?? false,
        };
        this.logger = new common_1.Logger(this.options.context);
    }
    debug(message, context) {
        if (this.shouldLog('debug')) {
            if (context && context !== this.options.context) {
                this.logger.debug(message, context);
            }
            else {
                this.logger.debug(message);
            }
        }
    }
    verbose(message, context) {
        if (this.shouldLog('verbose')) {
            if (context && context !== this.options.context) {
                this.logger.verbose(message, context);
            }
            else {
                this.logger.verbose(message);
            }
        }
    }
    log(message, context) {
        if (this.shouldLog('log')) {
            if (context && context !== this.options.context) {
                this.logger.log(message, context);
            }
            else {
                this.logger.log(message);
            }
        }
    }
    warn(message, context) {
        if (this.shouldLog('warn')) {
            if (context && context !== this.options.context) {
                this.logger.warn(message, context);
            }
            else {
                this.logger.warn(message);
            }
        }
    }
    error(message, error, context) {
        if (this.shouldLog('error') && this.options.logErrors) {
            const useContext = context && context !== this.options.context ? context : undefined;
            if (error instanceof Error) {
                if (useContext) {
                    this.logger.error(message, error.stack, useContext);
                }
                else {
                    this.logger.error(message, error.stack);
                }
            }
            else {
                if (useContext) {
                    this.logger.error(message, error, useContext);
                }
                else {
                    this.logger.error(message, error);
                }
            }
        }
    }
    performance(message, duration, context) {
        if (this.options.logPerformance && this.shouldLog('verbose')) {
            if (context && context !== this.options.context) {
                this.logger.verbose(`${message} (${duration}ms)`, context);
            }
            else {
                this.logger.verbose(`${message} (${duration}ms)`);
            }
        }
    }
    detail(message, data, context) {
        if (this.options.logDetails && this.shouldLog('debug')) {
            const useContext = context && context !== this.options.context ? context : undefined;
            if (data) {
                const detailMessage = `${message}: ${JSON.stringify(data, null, 2)}`;
                if (useContext) {
                    this.logger.debug(detailMessage, useContext);
                }
                else {
                    this.logger.debug(detailMessage);
                }
            }
            else {
                if (useContext) {
                    this.logger.debug(message, useContext);
                }
                else {
                    this.logger.debug(message);
                }
            }
        }
    }
    lifecycle(event, details, context) {
        const message = details ? `${event} ${JSON.stringify(details)}` : event;
        this.log(message, context);
    }
    methodCall(method, service, duration, context) {
        const baseMessage = `Method call: ${service}.${method}`;
        const message = duration ? `${baseMessage} (${duration}ms)` : baseMessage;
        if (duration && this.options.logPerformance) {
            this.verbose(message, context);
        }
        else {
            this.debug(message, context);
        }
    }
    connection(event, target, details, context) {
        const message = details
            ? `${event} to ${target} ${JSON.stringify(details)}`
            : `${event} to ${target}`;
        this.log(message, context);
    }
    shouldLog(level) {
        if (!this.options.enabled) {
            return false;
        }
        const levels = ['debug', 'verbose', 'log', 'warn', 'error'];
        const currentLevelIndex = levels.indexOf(this.options.level);
        const targetLevelIndex = levels.indexOf(level);
        return targetLevelIndex >= currentLevelIndex;
    }
    child(context) {
        return new GrpcLogger({
            ...this.options,
            context: `${this.options.context}:${context}`,
        });
    }
}
exports.GrpcLogger = GrpcLogger;
