// nestjs-grpc/dist/utils/proto-utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadProto = loadProto;
exports.loadProtoWithProtobuf = loadProtoWithProtobuf;
exports.getServiceByName = getServiceByName;
exports.getPackageByName = getPackageByName;
exports.getServiceMethods = getServiceMethods;
exports.createClientCredentials = createClientCredentials;
exports.createChannelOptions = createChannelOptions;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const grpc = __importStar(require("@grpc/grpc-js"));
const protoLoader = __importStar(require("@grpc/proto-loader"));
const protobuf = __importStar(require("protobufjs"));
const defaultOptions = {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true,
    arrays: true,
    objects: true,
    includeDirs: [],
};
function validateProtoPath(protoPath) {
    if (!protoPath || typeof protoPath !== 'string') {
        throw new Error('Proto path must be a non-empty string');
    }
    const resolvedPath = path.resolve(protoPath.trim());
    try {
        fs.accessSync(resolvedPath, fs.constants.R_OK);
    }
    catch {
        throw new Error(`Proto file not accessible: ${resolvedPath}`);
    }
    const stats = fs.statSync(resolvedPath);
    if (!stats.isFile()) {
        throw new Error(`Proto path is not a file: ${resolvedPath}`);
    }
    return resolvedPath;
}
function validateOptions(options) {
    if (typeof options !== 'object' || options === null) {
        throw new Error('Options must be an object');
    }
    const mergedOptions = {
        ...defaultOptions,
        ...options,
    };
    if (mergedOptions.includeDirs && !Array.isArray(mergedOptions.includeDirs)) {
        throw new Error('includeDirs must be an array');
    }
    return mergedOptions;
}
async function loadProto(protoPath, options = {}) {
    try {
        const validatedPath = validateProtoPath(protoPath);
        const validatedOptions = validateOptions(options);
        const packageDefinition = await protoLoader.load(validatedPath, validatedOptions);
        if (!packageDefinition) {
            throw new Error('Failed to load package definition from proto file');
        }
        const grpcObject = grpc.loadPackageDefinition(packageDefinition);
        if (!grpcObject || typeof grpcObject !== 'object') {
            throw new Error('Failed to create gRPC object from package definition');
        }
        return grpcObject;
    }
    catch (error) {
        if (error.message.includes('ENOENT')) {
            throw new Error(`Proto file not found: ${protoPath}`);
        }
        else if (error.message.includes('parse')) {
            throw new Error(`Proto file parse error: ${error.message}`);
        }
        else if (error.message.includes('EACCES')) {
            throw new Error(`Permission denied accessing proto file: ${protoPath}`);
        }
        else {
            throw new Error(`Failed to load proto file: ${error.message}`);
        }
    }
}
async function loadProtoWithProtobuf(protoPath) {
    try {
        const validatedPath = validateProtoPath(protoPath);
        const root = await protobuf.load(validatedPath);
        if (!root) {
            throw new Error('Protobufjs returned null root');
        }
        return root;
    }
    catch (error) {
        if (error.message.includes('ENOENT')) {
            throw new Error(`Proto file not found: ${protoPath}`);
        }
        else if (error.message.includes('parse')) {
            throw new Error(`Proto file parse error: ${error.message}`);
        }
        else {
            throw new Error(`Failed to load proto with protobufjs: ${error.message}`);
        }
    }
}
function getServiceByName(packageDefinition, packageName, serviceName) {
    try {
        validatePackageDefinition(packageDefinition);
        validatePackageName(packageName);
        validateServiceName(serviceName);
        const pkg = getPackageByName(packageDefinition, packageName);
        if (!pkg) {
            throw new Error(`Package '${packageName}' not found`);
        }
        if (!pkg[serviceName]) {
            const availableServices = Object.keys(pkg).filter(key => typeof pkg[key] === 'function');
            throw new Error(`Service '${serviceName}' not found in package '${packageName}'. ` +
                `Available services: ${availableServices.join(', ')}`);
        }
        const service = pkg[serviceName];
        if (typeof service !== 'function') {
            throw new Error(`'${serviceName}' is not a valid service constructor`);
        }
        return service;
    }
    catch (error) {
        throw new Error(`Failed to get service: ${error.message}`);
    }
}
function validatePackageDefinition(packageDefinition) {
    if (!packageDefinition || typeof packageDefinition !== 'object') {
        throw new Error('Package definition must be a valid object');
    }
}
function validatePackageName(packageName) {
    if (!packageName || typeof packageName !== 'string') {
        throw new Error('Package name must be a non-empty string');
    }
    if (packageName.trim().length === 0) {
        throw new Error('Package name cannot be empty');
    }
    const validPackageNameRegex = /^[a-zA-Z][a-zA-Z0-9_.]*$/;
    if (!validPackageNameRegex.test(packageName.trim())) {
        throw new Error('Package name contains invalid characters');
    }
}
function validateServiceName(serviceName) {
    if (!serviceName || typeof serviceName !== 'string') {
        throw new Error('Service name must be a non-empty string');
    }
    if (serviceName.trim().length === 0) {
        throw new Error('Service name cannot be empty');
    }
    const validServiceNameRegex = /^[a-zA-Z][a-zA-Z0-9_]*$/;
    if (!validServiceNameRegex.test(serviceName.trim())) {
        throw new Error('Service name contains invalid characters');
    }
}
function getPackageByName(packageDefinition, packageName) {
    try {
        validatePackageDefinition(packageDefinition);
        if (!packageName || typeof packageName !== 'string') {
            return packageDefinition;
        }
        const trimmedPackageName = packageName.trim();
        if (trimmedPackageName.length === 0) {
            return packageDefinition;
        }
        const parts = trimmedPackageName.split('.');
        let current = packageDefinition;
        for (const part of parts) {
            if (!part || part.trim().length === 0) {
                throw new Error('Package name contains empty parts');
            }
            if (!current || typeof current !== 'object') {
                throw new Error(`Invalid package structure at '${part}'`);
            }
            if (!current[part]) {
                return null;
            }
            current = current[part];
        }
        return current;
    }
    catch (error) {
        throw new Error(`Failed to find package '${packageName}': ${error.message}`);
    }
}
function getServiceMethods(serviceConstructor) {
    try {
        if (!serviceConstructor) {
            throw new Error('Service constructor is required');
        }
        if (typeof serviceConstructor !== 'function') {
            throw new Error('Service constructor must be a function');
        }
        if (!serviceConstructor.service) {
            return [];
        }
        const service = serviceConstructor.service;
        let methods = [];
        if (service.originalName && typeof service.originalName === 'object') {
            methods = Object.keys(service.originalName);
        }
        else if (service.methods && typeof service.methods === 'object') {
            methods = Object.keys(service.methods);
        }
        else if (service.methodsMap && typeof service.methodsMap === 'object') {
            methods = Object.keys(service.methodsMap);
        }
        else if (service.methods && Array.isArray(service.methods)) {
            methods = service.methods.map((method) => method.name).filter(Boolean);
        }
        else {
            methods = Object.keys(service).filter(key => key !== 'service' && key !== 'constructor' && typeof service[key] === 'object');
        }
        const validMethods = methods.filter(method => method && typeof method === 'string' && method.trim().length > 0);
        return validMethods;
    }
    catch (_error) {
        return [];
    }
}
function createClientCredentials(secure = false, rootCerts, privateKey, certChain) {
    try {
        if (!secure) {
            return grpc.credentials.createInsecure();
        }
        if (rootCerts && !Buffer.isBuffer(rootCerts)) {
            throw new Error('rootCerts must be a Buffer');
        }
        if (privateKey && !Buffer.isBuffer(privateKey)) {
            throw new Error('privateKey must be a Buffer');
        }
        if (certChain && !Buffer.isBuffer(certChain)) {
            throw new Error('certChain must be a Buffer');
        }
        return grpc.credentials.createSsl(rootCerts ?? null, privateKey ?? null, certChain ?? null);
    }
    catch (error) {
        throw new Error(`Failed to create client credentials: ${error.message}`);
    }
}
function createChannelOptions(maxSendSize, maxReceiveSize, additionalOptions = {}) {
    try {
        if (maxSendSize !== undefined) {
            if (!Number.isInteger(maxSendSize) || maxSendSize <= 0) {
                throw new Error('maxSendSize must be a positive integer');
            }
        }
        if (maxReceiveSize !== undefined) {
            if (!Number.isInteger(maxReceiveSize) || maxReceiveSize <= 0) {
                throw new Error('maxReceiveSize must be a positive integer');
            }
        }
        if (additionalOptions && typeof additionalOptions !== 'object') {
            throw new Error('additionalOptions must be an object');
        }
        const options = {
            'grpc.keepalive_time_ms': 60000,
            'grpc.keepalive_timeout_ms': 20000,
            'grpc.http2.min_time_between_pings_ms': 60000,
            'grpc.http2.max_pings_without_data': 0,
            'grpc.keepalive_permit_without_calls': 1,
        };
        if (maxSendSize) {
            options['grpc.max_send_message_length'] = maxSendSize;
        }
        if (maxReceiveSize) {
            options['grpc.max_receive_message_length'] = maxReceiveSize;
        }
        const safeAdditionalOptions = {};
        for (const [key, value] of Object.entries(additionalOptions || {})) {
            if (typeof key === 'string' && key.trim().length > 0) {
                safeAdditionalOptions[key] = value;
            }
        }
        return {
            ...options,
            ...safeAdditionalOptions,
        };
    }
    catch (error) {
        throw new Error(`Failed to create channel options: ${error.message}`);
    }
}
