// nestjs-grpc/dist/services/grpc-registry.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrpcRegistryService = void 0;
const common_1 = require("@nestjs/common");
const logger_1 = require("../utils/logger");
const grpc_provider_service_1 = require("./grpc-provider.service");
let GrpcRegistryService = class GrpcRegistryService {
    constructor(providerService) {
        this.providerService = providerService;
        this.logger = new logger_1.GrpcLogger({ context: 'GrpcRegistry' });
        this.pendingRegistrations = new Map();
        this.isProcessingRegistrations = false;
    }
    onModuleInit() {
        this.logger.lifecycle('Processing pending controller registrations');
        this.processPendingRegistrations();
    }
    registerController(serviceName, instance, metadata) {
        try {
            this.logger.debug(`Registering controller: ${serviceName}`);
            this.pendingRegistrations.set(serviceName, { instance, metadata });
            this.providerService.registerController(serviceName, instance, metadata);
        }
        catch (error) {
            this.logger.error(`Failed to register controller ${serviceName}`, error);
            throw error;
        }
    }
    getRegisteredControllers() {
        return new Map(this.pendingRegistrations);
    }
    isControllerRegistered(serviceName) {
        return this.pendingRegistrations.has(serviceName);
    }
    processPendingRegistrations() {
        if (this.isProcessingRegistrations || this.pendingRegistrations.size === 0) {
            return;
        }
        this.isProcessingRegistrations = true;
        try {
            const registrations = Array.from(this.pendingRegistrations.entries());
            let successCount = 0;
            let errorCount = 0;
            for (const [serviceName, { instance, metadata }] of registrations) {
                try {
                    this.providerService.registerController(serviceName, instance, metadata);
                    successCount++;
                }
                catch (error) {
                    this.logger.error(`Failed to register pending controller ${serviceName}`, error);
                    errorCount++;
                }
            }
            this.logger.lifecycle('Pending registrations processed', {
                total: registrations.length,
                successful: successCount,
                failed: errorCount,
            });
            this.pendingRegistrations.clear();
        }
        finally {
            this.isProcessingRegistrations = false;
        }
    }
    async waitForProviderReady() {
        let attempts = 0;
        const maxAttempts = 10;
        const delay = 100;
        while (attempts < maxAttempts) {
            if (this.providerService.isServerRunning()) {
                return;
            }
            this.logger.debug(`Provider not ready, attempt ${attempts + 1}/${maxAttempts}`);
            await new Promise(resolve => setTimeout(resolve, delay));
            attempts++;
        }
        this.logger.warn('Provider service not ready after maximum attempts, proceeding anyway');
    }
};
exports.GrpcRegistryService = GrpcRegistryService;
exports.GrpcRegistryService = GrpcRegistryService = __decorate([
    (0, common_1.Injectable)(),
    __metadata("design:paramtypes", [grpc_provider_service_1.GrpcProviderService])
], GrpcRegistryService);
